collectors/node/network.go

```diff
@@ -0,0 +1,153 @@
+	netOutPackets uint64
+	netInDropped  uint64
+	netOutDropped uint64
+	netInErrors   uint64
+	netOutErrors  uint64
+	timestamp     string
+}
+
+func (m *networkMetrics) poll() error {
+	netMetrics := []networkMetric{}
+
+	ts := thisTime()
+	netInterface, err := net.IOCounters(true) // per network interface
+	if err != nil {
+		return err
+	}
+
+	for _, nic := range netInterface {
+		netMetrics = append(netMetrics, networkMetric{
+			interfaceName: nic.Name,
+			netIn:         nic.BytesRecv,
+			netOut:        nic.BytesSent,
+			netInPackets:  nic.PacketsRecv,
+			netOutPackets: nic.PacketsSent,
+			netInDropped:  nic.Dropin,
+			netOutDropped: nic.Dropout,
+			netInErrors:   nic.Errin,
+			netOutErrors:  nic.Errout,
+			timestamp:     ts,
+		})
+	}
+
+	m.interfaces = netMetrics
+
+	return nil
+}
+
+func (m *networkMetrics) addDatapoints(nc *nodeCollector) error {
+
+	for _, nic := range m.interfaces {
+		nc.datapoints = append(nc.datapoints, producers.Datapoint{
+			Name:      NET_IN,
+			Unit:      BYTES,
+			Value:     nic.netIn,
+			Timestamp: nic.timestamp,
+			Tags: map[string]string{
+				"interface": nic.interfaceName,
+			},
+		})
+
+		nc.datapoints = append(nc.datapoints, producers.Datapoint{
+			Name:      NET_OUT,
+			Unit:      BYTES,
+			Value:     nic.netOut,
+			Timestamp: nic.timestamp,
+			Tags: map[string]string{
+				"interface": nic.interfaceName,
+			},
+		})
+
+		nc.datapoints = append(nc.datapoints, producers.Datapoint{
+			Name:      NET_IN_PACKETS,
+			Unit:      COUNT,
+			Value:     nic.netInPackets,
+			Timestamp: nic.timestamp,
+			Tags: map[string]string{
+				"interface": nic.interfaceName,
+			},
+		})
+
+		nc.datapoints = append(nc.datapoints, producers.Datapoint{
+			Name:      NET_OUT_PACKETS,
+			Unit:      COUNT,
+			Value:     nic.netOutPackets,
+			Timestamp: nic.timestamp,
+			Tags: map[string]string{
+				"interface": nic.interfaceName,
+			},
+		})
+
+		nc.datapoints = append(nc.datapoints, producers.Datapoint{
+			Name:      NET_IN_DROPPED,
+			Unit:      COUNT,
+			Value:     nic.netInDropped,
+			Timestamp: nic.timestamp,
+			Tags: map[string]string{
+				"interface": nic.interfaceName,
+			},
+		})
+
+		nc.datapoints = append(nc.datapoints, producers.Datapoint{
+			Name:      NET_OUT_DROPPED,
+			Unit:      COUNT,
+			Value:     nic.netOutDropped,
+			Timestamp: nic.timestamp,
+			Tags: map[string]string{
+				"interface": nic.interfaceName,
+			},
+		})
+
+		nc.datapoints = append(nc.datapoints, producers.Datapoint{
+			Name:      NET_IN_ERRORS,
+			Unit:      COUNT,
+			Value:     nic.netInErrors,
+			Timestamp: nic.timestamp,
+			Tags: map[string]string{
+				"interface": nic.interfaceName,
+			},
+		})
+
+		nc.datapoints = append(nc.datapoints, producers.Datapoint{
+			Name:      NET_OUT_ERRORS,
+			Unit:      COUNT,
+			Value:     nic.netOutErrors,
+			Timestamp: nic.timestamp,
+			Tags: map[string]string{
+				"interface": nic.interfaceName,
+			},
+		})
+	}
+
+	return nil
+}
```

Comment:
i have the same comments here, reuse the structure, use lock.Mutax, do not update reference to a structure.